### DISCUSSION TITLE
LSP Server in Lualine

### DISCUSSION THREAD

**Discussion by karimlevallois** [LazyVim/LazyVim#2595] [Q&A]
Hi everyone!

Sorry for all the questions, moving from LunarVim.

Does anyone know how to add the LSP servers, Pyright, Ruff, Black etc.  on the right of lualine? Basically I want my lualine to look like the below: -

<img width="1680" alt="Screenshot 2024-02-22 at 19 24 14" src="https://github.com/LazyVim/LazyVim/assets/80129273/944257ee-3828-4fc6-a8d2-4552761da2da">

Many thanks in advance.

**Comment by dpetka2001** [comment]
See #2172.

**Reply by karimlevallois** [reply]
Awesome, got that working!

Lastly, if I wanted the right section to match LunarVim with lsp, indentation and filetype/fileformat (not sure which it would be) along with straight separators where would I out this options? Currently I have the following at the bottom of my lsp/lualine file: -

```
return {
  {
    "nvim-lualine/lualine.nvim",
    opts = function(_, opts)
      table.insert(opts.sections.lualine_x, lsp)
    end,
  },
}
```

**Reply by dpetka2001** [reply]
No idea. I'm not using this kind of config. Try to play around with it to see what happens.

**Reply by karimlevallois** [reply]
Thanks for all your patience, it's much appreciated.

I managed to get lualine looking as I like with the following: -

```
local lsp = {
  function()
    local buf_clients = vim.lsp.get_active_clients({ bufnr = 0 })
    if #buf_clients == 0 then
      return "LSP Inactive"
    end

    local formatters = require("conform").list_formatters(0)
    local linters = require("lint").linters_by_ft[vim.bo.filetype]

    local buf_client_names = {}
    local buf_formatters = {}
    local buf_linters = {}

    -- add client
    for _, client in pairs(buf_clients) do
      if client.name ~= "null-ls" and client.name ~= "copilot" then
        table.insert(buf_client_names, client.name)
      end
    end

    -- add formatter
    for _, formatter in pairs(formatters) do
      table.insert(buf_formatters, formatter.name)
    end

    -- add linter -- disabled as this breaks things
    -- for _, linter in pairs(linters) do
    --   table.insert(buf_linters, linter)
    -- end

    vim.list_extend(buf_client_names, buf_formatters)
    vim.list_extend(buf_client_names, buf_linters)

    local unique_client_names = table.concat(buf_client_names, ", ")
    local language_servers = string.format("[%s]", unique_client_names)

    return language_servers
  end,
  color = { gui = "bold" },
}

location = { "location" }
progress = {
  "progress",
  fmt = function()
    return "%P/%L"
  end,
  color = {},
}

spaces = {
  function()
    local shiftwidth = vim.api.nvim_buf_get_option(0, "shiftwidth")
    return "󰌒" .. " " .. shiftwidth
  end,
  padding = 1,
}

scrollbar = {
  function()
    local current_line = vim.fn.line(".")
    local total_lines = vim.fn.line("$")
    local chars = { "__", "▁▁", "▂▂", "▃▃", "▄▄", "▅▅", "▆▆", "▇▇", "██" }
    local line_ratio = current_line / total_lines
    local index = math.ceil(line_ratio * #chars)
    return chars[index]
  end,
  padding = { left = 0, right = 0 },
  color = "SLProgress",
  cond = nil,
}

return {
  {
    "nvim-lualine/lualine.nvim",
    opts = function(_, opts)
      table.insert(opts.sections.lualine_x, lsp)
    end,

    {
      "nvim-lualine/lualine.nvim",
      event = "VeryLazy",
      opts = {
        options = {
          component_separators = "",
          section_separators = "",
        },
        sections = {
          lualine_x = { lsp, spaces },
          lualine_y = { location },
          lualine_z = { progress },
        },
        inactive_sections = {
          lualine_x = { lsp, spaces },
          lualine_y = { location },
          lualine_z = { progress },
        },
      },
    },
  },
}
```

There is one funny thing with the lsp servers though, the ordering is never consistent, sometimes it will show ruff_lsp, pyright, black and sometime pyright, ruff_lsp and black. Is there any way to show them consistently, so always pyright, black and ruff_lsp - so lsp, formatter then linter?

**Reply by dpetka2001** [reply]
I'm not sure if it's possible with another method, but this method utilizes `pairs` function to iterate over the active clients and `pairs` function always returns items in arbitrary order, so never consistent. And even the function `vim.lsp.get_active_clients` that is used for retrieving the attached clients on the buffer, uses `pairs` under the hood.

**Reply by karimlevallois** [reply]
Strange, wonder how LunarVim does it, will have to dig deeper.

**Reply by karimlevallois** [reply]
At least I can get in some order with...

```
    table.sort(buf_client_names)
    local unique_client_names = table.concat(buf_client_names, ", ")
    local language_servers = string.format("[%s]", unique_client_names)
```

**Comment by karimlevallois** [comment]
So, I cannot seem to fix the linter issue when using ruff_lsp as a 'linter' (through Lazy Extras - Python) for example. When running python files neither client, formatter or linter is showing (just completely blank) when I have the below enabled. I think it breaks when no linter is applied to the file type, so returns nothing at all for anything.

Is there a way to set a fallback, so if no linter is applied, it just returns the client and formatter?

 ```
    for _, linter in pairs(linters) do
      table.insert(buf_linters, linter)
    end
```
